src/cppkin.h

```diff
@@ -14,87 +14,96 @@
-            outputData = std::move(context.ToString()); \
-        } \
-        else \
-            outputData = std::string(); \
-    }while(0)
-
-#define DESERIALIZE_TRACING_HEADER(decodingType, data, length, outputHeader) \
-    do  \
-    {   \
-        if(length > 0){ \
-            cppkin::EncoderContext##decodingType context(data, length); \
-            outputHeader = cppkin::Encoder<EncodingType::decodingType>::DeSerializeSpanHeader(context); \
-        } \
-    }while(0)
-
-#define SUBMIT_SPAN()  \
-    do  \
-    {   \
-        if(cppkin::Sampler::ShouldSample()){ \
-            std::unique_ptr<cppkin::Span> span = cppkin::SpanContainer::Instance().ReleaseSpan();   \
-            span->SetEndTime(); \
-            cppkin::TransportManager::Instance().PushSpan(std::move(span)); \
-        } \
-    }while(0)
-
-#define TRACE_EVENT(eventName) \
-    do  \
-    {   \
-        if(cppkin::Sampler::ShouldSample()){ \
-            cppkin::Span& span = cppkin::SpanContainer::Instance().GetSpan();   \
-            span.CreateSimpleAnnotation(eventName); \
-        } \
-    }while(0)
+    static inline void CreateTrace(const char* operationName){
+        if(!cppkin::Sampler::AdvanceSampler()) {
+            cppkin::SpanContainer::Instance().ResetRootHeader();
+            return;
+        }
+        uint_fast64_t id = cppkin::Trace::Instance().GenerateTraceID();
+        std::unique_ptr<cppkin::Span> span = std::move(cppkin::Trace::Instance().CreateSpan(operationName, id));
+        cppkin::SpanContainer::Instance().PushSpan(std::move(span));
+    }
+
+    static inline void CreateSpan(const char* operationName){
+
+        if (!cppkin::Sampler::ShouldSample())
+            return;
+
+        cppkin::SpanContainer& spanContainer = cppkin::SpanContainer::Instance();
+        const cppkin::Span::SpanHeader* parentHeader = spanContainer.GetRootHeader();
+        if(!parentHeader) // No parent available. Span always has to have a parent. Skipping
+            return;
+
+        const cppkin::Span* currentSpan = spanContainer.TopSpan();
+        if(currentSpan) {
+            parentHeader = &currentSpan->GetHeader();
+        }
+
+        uint_fast64_t id = cppkin::Trace::Instance().GenerateSpanID();
+        std::unique_ptr<cppkin::Span> span = cppkin::Trace::Instance().CreateSpan(operationName, parentHeader->TraceID, parentHeader->ID, id);
+        spanContainer.PushSpan(std::move(span));
+    }
+
+    static inline void JoinSpan(const cppkin::Span::SpanHeader& spanHeader) {
+
+        if (!cppkin::Sampler::ShouldSample())
+            return;
+
+        std::unique_ptr<cppkin::Span> span;
+        if(spanHeader.ParentIDSet)
+             span = std::move(cppkin::Trace::Instance().CreateSpan(spanHeader.Name.c_str(), spanHeader.TraceID, spanHeader.ParentID, spanHeader.ID));
+        else
+            span = std::move(cppkin::Trace::Instance().CreateSpan(spanHeader.Name.c_str(), spanHeader.TraceID));
```

Comment:
When joining a trace, should we set the tls sample count to be valid or not according to what it used to be in prrvious service?